client/rest/src/plugins/metadata/metal.js

```diff
@@ -2,59 +2,57 @@ const catapult = require('../../catapult-sdk');
 const { sha3_256 } = require('@noble/hashes/sha3');
 const bs58 = require('bs58');
 
-const METAL_ID_HEADER_HEX = '0B2A';
-const METAL_ID_LENGTH = 68;
+const METAL_ID_HEADER_HEX = [0x0B, 0x2A];
+const METAL_ID_LENGTH = 34;
 
 const metal = {
 	/**
-	 * convert MetalID to compositehash
-	 * @param {string} metalId ex) FeF65JftVPEGwaua35LnbU9jK46uG3W8karGDDuDwVEh8Z.
-	 * @returns {string} composite hash.
+	 * Convert a metal id (e.g. 'FeF65JftVPEGwaua35LnbU9jK46uG3W8karGDDuDwVEh8Z') into its associated composite hash.
+	 * @param {string} metalId Metal id to convert.
+	 * @returns {Uint8Array} Associated composite hash.
 	 */
-	restoreMetadataHash(metalId) {
-		const hashHex = catapult.utils.convert.uint8ToHex(bs58.decode(metalId));
-		const isValidMetalId = hashHex.startsWith(METAL_ID_HEADER_HEX) && METAL_ID_LENGTH === hashHex.length;
+	extractCompositeHashFromMetalId(metalId) {
+		const decodedBs58HashBytes = bs58.decode(metalId);
+		const isValidMetalId = METAL_ID_HEADER_HEX[0] === decodedBs58HashBytes[0]
+			&& METAL_ID_HEADER_HEX[1] === decodedBs58HashBytes[1]
+			&& METAL_ID_LENGTH === decodedBs58HashBytes.length;
 
 		if (!isValidMetalId)
-			throw Error('Invalid metal ID.');
+			throw Error(`'${metalId}' is not a valid metal id`);
 
-		const compositeHash = hashHex.slice(METAL_ID_HEADER_HEX.length);
-		return { compositeHash };
+		return decodedBs58HashBytes
+			.subarray(METAL_ID_HEADER_HEX.length);
 	},
 
 	/**
 	 * generate metadata key for checking the metadata is broken or not.
 	 * @param {Uint8Array} input value of chunk.
-	 * @returns {Uint32Array} Return Metadata key as Uint32Array.
+	 * @returns {module:utils/uint64} Return Metadata key as uint64.
 	 */
 	generateMetadataKey(input) {
 		if (!input.length)
-			throw new Error('Input must not be empty');
-		const { buffer } = sha3_256(input);
-		const uint32Array = new Uint32Array(buffer);
-		return [uint32Array[0], uint32Array[1] & 0x7FFFFFFF];
+			throw new Error('input must not be empty');
+		const hashedBytes = sha3_256(input);
+		const uint64 = catapult.utils.uint64.fromBytes(hashedBytes.subarray(0, 8));
+		return [uint64[0], uint64[1] & 0x7FFFFFFF];
 	},
 
 	/**
 	 * extract chunk and return the value of a metadata entry according to the metadata protocol.
-	 * @param {Buffer} b Value obtained from metadata entry.
+	 * @param {Buffer} buffer Value obtained from metadata entry.
 	 * @returns {object} Objects containing values in line with the METAL protocol.
 	 * - magic: number
 	 * - text: boolean
 	 * - scopedMetadataKey: uint64
 	 * - chunkPayload: Buffer
 	 */
```

Comment:
I'm not too sure about this one.
what do you think?
```js
/**
* generates a metadata key to check if the metadata is broken or not.
* @param {Uint8Array} input The chunk value to generate the key from.
* @returns {module:utils/uint64} The metadata key as a uint64, represented as an array of two numbers.
*/
generateMetadataKey(input) {
if (!input.length)
	throw new Error('input must not be empty');
const hashedBytes = sha3_256(input);
const uint64 = catapult.utils.uint64.fromBytes(hashedBytes.subarray(0, 8));
return [uint64[0], uint64[1] & 0x7FFFFFFF];
},
```

Reply:
how do you think?
```js
/**
* extract chunk and return the value of a metadata entry according to the metadata protocol.
* @param {Buffer} buffer Value obtained from metadata entry.
* @returns {object} Objects containing values in line with the METAL protocol.
* - magic: number - 0x00 is chunk, 0x80 is end of chunk.
* - text: boolean - true if text is present, false otherwise.
* - scopedMetadataKey: uint64 - ScopedMetadataKey of the next chunk.
* - chunkPayload: Buffer - Payload of the chunk.
*/
extractChunk(buffer) {
const header = buffer.subarray(0, 1);
const magic = header[0] & 0x80;
const text = Boolean(header[0] & 0x40);

const scopedMetadataKey = catapult.utils.uint64.fromBytes(new Uint8Array(buffer.subarray(4, 12)).reverse());
const chunkPayload = buffer.subarray(12);

return {
	magic, text, scopedMetadataKey, chunkPayload
};
},
```